astropy/io/fits/file.py

```diff
@@ -119,9 +120,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         if mode not in IO_FITS_MODES:
             raise ValueError("Mode '{}' not recognized".format(mode))
 
+        # Handle raw URLs
         if (isinstance(fileobj, string_types) and
-            mode not in ('ostream', 'append') and
-            _is_url(fileobj)):  # This is an URL.
+            mode not in ('ostream', 'append', 'update') and _is_url(fileobj)):
+            self.name = download_file(fileobj, cache=cache)
+        # Handle responses from URL requests that have already been opened
+        elif isinstance(fileobj, http_client.HTTPResponse):
```

Comment:
Also it does seem like a good idea handle `Request` as well, but I don't think it's as simple as checking for `hasattr(fileobj, 'url')` since I think the `Request` attribute is `full_url`. It will still be necessary to test for `isinstance(..., HTTPResponse)` and `isinstance(..., Request)` since they will need to be handled differently.